### ISSUE TITLE
feat: Assignment document incremental history using JSON Patch

### ISSUE THREAD

**Issue by armorup** [codepetca/pika#83]
## Summary

Implement efficient incremental history tracking for assignment documents using JSON Patch (RFC 6902). Store diffs between versions instead of full snapshots to minimize storage costs while enabling version recovery, academic integrity auditing, and work-over-time analytics.

## Background / Context

Currently, each autosave overwrites `assignment_docs.content` with no history. This means:
- Students can't recover accidentally deleted work
- Teachers can't see how work evolved over time
- No audit trail for academic integrity disputes
- No data to distinguish "worked for 2 weeks" vs "copied night before"

**Related:**
- `docs/semester-plan.md` - Priority 2A (reliability + grading)
- PR #82 - Draft autosave (localStorage, complementary feature)
- Migration 012 - assignment_docs uses JSONB content

## Requirements / Acceptance Criteria

- [ ] Create `assignment_doc_history` table with patch storage
- [ ] Capture history entry on every autosave (PATCH to `/api/assignment-docs/[id]`)
- [ ] Capture history entry on editor blur (focus loss = natural breakpoint)
- [ ] Store FULL snapshot only on: (1) first edit, (2) submission
- [ ] JSON Patch format using `fast-json-patch` library
- [ ] History entries include: patch, timestamp, word_count, char_count
- [ ] Retention: auto-purge history 30 days after `classroom.end_date`
- [ ] Teacher can view document history timeline
- [ ] Student can restore previous version of their own work
- [ ] Storage budget: patches only during drafting (no redundant full saves)

## Scope

**In scope**
- Assignment documents (`assignment_docs` table) history tracking
- JSON Patch diff generation and storage
- History viewing UI (timeline with word count graph)
- Version restore functionality
- Semester-end cleanup job

**Out of scope**
- Daily journal entries (entries table) - different use case
- Real-time collaboration / conflict resolution
- AI-powered "significant change" detection
- Diff visualization (show exact changes highlighted)

### Database Schema

```sql
CREATE TABLE assignment_doc_history (
  id uuid PRIMARY KEY DEFAULT gen_random_uuid(),
  assignment_doc_id uuid NOT NULL REFERENCES assignment_docs(id) ON DELETE CASCADE,
  
  -- Patch data (NULL for full snapshots)
  patch jsonb,
  
  -- Full snapshot (NULL for patch entries, stored on first edit + submission)
  snapshot jsonb,
  
  -- Metadata for analytics
  word_count int NOT NULL,
  char_count int NOT NULL,
  
  -- What triggered this entry
  trigger text NOT NULL CHECK (trigger IN ('autosave', 'blur', 'submit', 'baseline')),
  
  created_at timestamptz NOT NULL DEFAULT now()
);

CREATE INDEX idx_doc_history_lookup 
  ON assignment_doc_history(assignment_doc_id, created_at DESC);

CREATE INDEX idx_doc_history_cleanup
  ON assignment_doc_history(created_at);
```

### Key Files
- `src/lib/json-patch.ts` - Patch generation/application utilities
- `src/app/api/assignment-docs/[id]/route.ts` - Add history capture on PATCH
- `src/app/api/assignment-docs/[id]/history/route.ts` - New: fetch history
- `src/app/api/assignment-docs/[id]/restore/route.ts` - New: restore version
- `src/app/api/cron/cleanup-history/route.ts` - Semester-end purge job

### Constraints
- Must work with existing Tiptap JSONB content structure
- Patches must be reversible (can reconstruct any version)
- Storage-efficient: patches ~100-500 bytes vs full docs ~5-10KB
- Rate limiting: at most 1 history entry per 10 seconds per doc

### Suggested Approach

1. **Baseline capture**: On first autosave after doc creation, store full snapshot with `trigger='baseline'`
2. **Patch capture**: Subsequent autosaves compute JSON Patch against last known state, store patch only
3. **Blur capture**: On editor blur event, capture current state (patch) with `trigger='blur'`
4. **Submit capture**: On submission, store full snapshot with `trigger='submit'`
5. **Reconstruction**: To rebuild version N, start from most recent snapshot before N, apply patches forward
6. **Cleanup cron**: Daily job deletes history where classroom ended 30+ days ago

### Automated
- [ ] Unit tests for JSON Patch generation/application (fast-json-patch)
- [ ] Unit tests for reconstruction logic (snapshot + patches → document)
- [ ] API tests for history capture on PATCH/blur/submit
- [ ] API tests for history retrieval
- [ ] API tests for version restore

### Manual
- [ ] Student edits assignment → history entries appear in DB
- [ ] Student blurs editor → blur entry captured
- [ ] Student submits → full snapshot captured
- [ ] Teacher views history timeline → shows entries with timestamps
- [ ] Student restores old version → content reverts correctly
- [ ] After classroom ends + 30 days → history purged

## Risks / Open Questions

1. **Patch size edge case**: What if a paste operation creates a huge patch? Consider: if patch > 50% of full doc size, store snapshot instead.

2. **Reconstruction performance**: With 100+ patches, reconstruction could be slow. Mitigation: compact patches into periodic snapshots via background job if needed (future optimization).

3. **Blur event reliability**: Browser blur events can be inconsistent. May need to also capture on visibility change or beforeunload.

4. **Race conditions**: Concurrent autosaves could create out-of-order patches. Mitigation: use optimistic locking or last-write-wins with timestamp ordering.

5. **Supabase storage limits**: Monitor storage growth. Patches should be ~25x more efficient than full snapshots, but worth tracking.
